src/crab/type_domain.cpp

```diff
@@ -0,0 +1,714 @@
+// Copyright (c) Prevail Verifier contributors.
+// SPDX-License-Identifier: MIT
+
+#include <unordered_map>
+
+#include "crab/type_domain.hpp"
+
+using crab::___print___;
+using crab::ptr_t;
+using crab::ptr_with_off_t;
+using crab::ptr_no_off_t;
+using crab::ctx_t;
+using crab::global_type_env_t;
+using crab::reg_with_loc_t;
+using crab::live_registers_t;
+using crab::register_types_t;
+
+static std::string size(int w) { return std::string("u") + std::to_string(w * 8); }
+
+
+namespace std {
+    template <>
+    struct hash<crab::reg_with_loc_t> {
+        size_t operator()(const crab::reg_with_loc_t& reg) const { return reg.hash(); }
+    };
+
+    // does not seem to work for me
+    template <>
+    struct equal_to<crab::ptr_t> {
+        constexpr bool operator()(const crab::ptr_t& p1, const crab::ptr_t& p2) const {
+            if (p1.index() != p2.index()) return false;
+            if (std::holds_alternative<crab::ptr_no_off_t>(p1)) {
+                auto ptr_no_off1 = std::get<crab::ptr_no_off_t>(p1);
+                auto ptr_no_off2 = std::get<crab::ptr_no_off_t>(p2);
+                return (ptr_no_off1.r == ptr_no_off2.r);
+            }
+            else {
+                auto ptr_with_off1 = std::get<crab::ptr_with_off_t>(p1);
+                auto ptr_with_off2 = std::get<crab::ptr_with_off_t>(p2);
+                return (ptr_with_off1.r == ptr_with_off2.r && ptr_with_off1.offset == ptr_with_off2.offset);
+            }
+        }
+    };
+}
+
+
+void print_ptr_type(const ptr_t& p) {
```

Comment:
Make it `static` since it's internal to this cpp file.

Reply:
ok